File: crates/qualification-driver/src/scenario/identity.rs
use serde::Deserialize;

use super::{ActionOutcome, Runner};

#[derive(Debug, Deserialize)]
struct StatusResponse {
    status: String,
    version: String,
    git_commit: String,
    environment: String,
}

/// The build context excludes version-control metadata, so an image built
/// without the commit passed in reports `unknown` and this assertion would
/// compare nothing.
pub async fn assert_identity(runner: &Runner) -> ActionOutcome {
    let url = format!("{}/status", runner.endpoints.http);
    let response = match runner.http.get(&url).send().await {
        Ok(response) => response,
        Err(error) => {
            return ActionOutcome::failed_setup(format!("GET {url} failed: {error}"));
        }
    };

    if !response.status().is_success() {
        return ActionOutcome::failed_product(format!(
            "GET {url} returned {}",
            response.status().as_u16()
        ));
    }

    let status: StatusResponse = match response.json().await {
        Ok(body) => body,
        Err(error) => {
            return ActionOutcome::failed_product(format!(
                "{url} returned an unreadable body: {error}"
            ));
        }
    };

    if status.status != "ok" {
        return ActionOutcome::failed_product(format!("server reports status `{}`", status.status));
    }

    if status.git_commit == "unknown" {
        return ActionOutcome::failed_setup(
            "server reports an unknown commit; the image was built without its source revision, \
             so the identity assertion would compare nothing"
                .to_string(),
        );
    }

    let expected = &runner.expectation.image_revision;
    if !expected.is_empty()
        && !expected.starts_with(&status.git_commit)
        && !status.git_commit.starts_with(expected.as_str())
    {
        return ActionOutcome::failed_setup(format!(
            "server reports commit `{}` but the artifact under test is `{expected}`",
            status.git_commit
        ));
    }

    if status.version.is_empty() || status.environment.is_empty() {
        return ActionOutcome::failed_product(
            "server reported an empty version or environment".to_string(),
        );
    }

    ActionOutcome::Passed
}

File: packages/miden-multisig-client/tests/qualification/actions/identity.ts
import type { ActionOutcome, ActionContext } from '../runner.js';

interface StatusResponse {
  status: string;
  version: string;
  git_commit: string;
  environment: string;
}

export async function assertIdentity(context: ActionContext): Promise<ActionOutcome> {
  const url = `${context.httpEndpoint}/status`;
  let response: Response;
  try {
    response = await fetch(url);
  } catch (error) {
    return { kind: 'failed', classification: 'setup', reason: `GET ${url} failed: ${String(error)}` };
  }

  if (!response.ok) {
    return { kind: 'failed', classification: 'product', reason: `GET ${url} returned ${response.status}` };
  }

  const status = (await response.json()) as StatusResponse;

  if (status.status !== 'ok') {
    return { kind: 'failed', classification: 'product', reason: `server reports status ${status.status}` };
  }
  if (status.git_commit === 'unknown') {
    return {
      kind: 'failed',
      classification: 'setup',
      reason:
        'server reports an unknown commit; the image was built without its source revision, so the identity assertion would compare nothing',
    };
  }
  const expected = context.imageRevision;
  if (expected && !expected.startsWith(status.git_commit) && !status.git_commit.startsWith(expected)) {
    return {
      kind: 'failed',
      classification: 'setup',
      reason: `server reports commit ${status.git_commit} but the artifact under test is ${expected}`,
    };
  }
  if (!status.version || !status.environment) {
    return { kind: 'failed', classification: 'product', reason: 'server reported an empty version or environment' };
  }
  return { kind: 'passed' };
}
